File: 362/solution.js
main();

function main() {
  expect("leetcode", ["leet", "code"], true);
  expect("catsandog", ["cat", "cats", "and", "sand", "dog"], false);
  expect("aaaaaaaa", ["aa", "aaa"], true);
}

function expect(s, dict, expected) {
  const actual = wordBreak(s, dict);
  if (actual != expected) {
    console.error(`Expected wordBreak([${s}], [${dict}]) == ${expected}, got ${actual}`);
  }
}

function wordBreak(s, dict) {
  return wordBreakHelper(s, new Set(dict));
}

function wordBreakHelper(s, dict) {
  let head = "";
  for (let i = 0; i < s.length; i++) {
    head += s[i];
    if (dict.has(head)) {
      const tail = s.substring(i+1);
      if (tail == "") return true;
      if (wordBreakHelper(tail, dict)) return true;
    }
  }
  return false;
}

File: 362/solution.go
package main

import (
	"fmt"
)

func main() {
	expect("leetcode", []string{"leet", "code"}, true)
	expect("catsandog", []string{"cat", "cats", "and", "sand", "dog"}, false)
	expect("aaaaaaaa", []string{"aa", "aaa"}, true)
}

func expect(s string, dict []string, expected bool) {
	actual := wordBreak(s, dict)
	if actual != expected {
		fmt.Printf("Expected wordBreak(%v, %v) == %v, got %v\n", s, dict, expected, actual)
	}
}

func wordBreak(s string, dict []string) bool {
	dictMap := make(map[string]bool)
	for _, word := range dict {
		dictMap[word] = true
	}
	return wordBreakHelper(s, dictMap)
}

func wordBreakHelper(s string, dict map[string]bool) bool {
	var head, tail string
	for i := range s {
		head = s[:i+1]
		if _, ok := dict[head]; ok {
			tail = s[i+1:]
			if tail == "" {
				return true
			}
			if wordBreakHelper(tail, dict) {
				return true
			}
		}
	}
	return false
}
